lib/inputs/pixel-trackers.js

```diff
@@ -0,0 +1,107 @@
+const crypto = require('crypto')
+const logger = require('../logger')
+const timestamp = require('../timestamp')
+const lookip = require('../lookip')
+const bigquery = require('../bigquery')
+
+/**
+ * Pixel.prx.org trackers
+ */
+module.exports = class PixelTrackers {
+
+  constructor(records) {
+    this._records = (records || []).filter(r => this.check(r))
+  }
+
+  check(record) {
+    return record.type === 'pixel'
+  }
+
+  async insert() {
+    if (this._records.length === 0) {
+      return []
+    }
+
+    // format records and organize by dataset + table
+    const formatted = await Promise.all(this._records.map(r => this.format(r)))
+    const grouped = this._records.reduce((acc, rec, idx) => {
+      const dest = this.destination(rec)
+      if (dest) {
+        (acc[dest] = acc[dest] || []).push(formatted[idx])
+      }
+      return acc
+    }, {})
+
+    // insert in parallel
+    return await Promise.all(Object.keys(grouped).map(async (datasetTable) => {
+      try {
+        const [ds, tbl] = datasetTable.split('.')
+        const num = await bigquery.insert(ds, tbl, grouped[datasetTable])
+        return {count: num, dest: datasetTable}
+      } catch (err) {
+        if (err.code === 404) {
+          logger.error(`Table not found: ${datasetTable}`)
+          return {count: 0, dest: datasetTable}
+        } else {
+          const insertErrors = logger.combineErrors(err)
+          if (insertErrors) {
+            logger.error(`Insert errors on ${datasetTable}: ${insertErrors}`)
+            return {count: 0, dest: datasetTable}
+          } else {
+            throw err
+          }
+        }
+      }
+    }))
+  }
+
+  async format(record) {
+    const epoch = timestamp.toEpochSeconds(record.timestamp || 0)
+    const userId = this.userId(record)
+    const geo = await lookip.look(record.remoteIp)
+    return {
+      insertId: this.insertId(epoch, userId, record.key, record.canonical),
+      json: {
+        timestamp:          epoch,
+        user_id:            userId,
+        key:                record.key,
+        canonical:          record.canonical,
+        remote_agent:       record.remoteAgent,
+        remote_referrer:    record.remoteReferrer,
```

Comment:
Are we going to be able to easily group these by host domain or something, so we could show Rick an aggregated number of referral traffic for all of wired.com or something like that?

Reply:
Yep, can do that with `net.host` in BigQuery ... dovetail example:

```
select net.host(remote_referrer), count(*) as count
from production.dt_downloads
where remote_referrer is not null and remote_referrer != '' and timestamp >= '2019-05-10'
group by remote_referrer
order by count desc
```